nengo_loihi/simulator.py

```diff
@@ -365,26 +365,34 @@ def _collect_receiver_info(self):
         spikes = []
         errors = OrderedDict()
         for sender, receiver in self.model.host2chip_senders.items():
-            receiver.clear()
+            spike_target = receiver.spike_target
+            error_target = receiver.error_target
+            if error_target is not None:
+                conn = self.model.probe_conns[error_target]
+                error_synapse = self.model.objs[conn]["decoders"]
+                assert error_synapse.learning
+            else:
+                error_synapse = None
+
             for t, x in sender.queue:
-                receiver.receive(t, x)
-            del sender.queue[:]
-
-            if hasattr(receiver, "collect_spikes"):
-                for spike_input, t, spike_idxs in receiver.collect_spikes():
-                    ti = round(t / self.model.dt)
-                    spikes.append((spike_input, ti, spike_idxs))
-            if hasattr(receiver, "collect_errors"):
-                for probe, t, e in receiver.collect_errors():
-                    conn = self.model.probe_conns[probe]
-                    synapse = self.model.objs[conn]["decoders"]
-                    assert synapse.learning
-                    ti = round(t / self.model.dt)
-                    errors_ti = errors.setdefault(ti, OrderedDict())
-                    if synapse in errors_ti:
-                        errors_ti[synapse] += e
+                ti = round(t / self.model.dt)
+
+                if spike_target is not None:
+                    spike_idxs = x.nonzero()[0]
+                    spikes.append((spike_target, ti, spike_idxs))
+
+                if error_synapse is not None:
+                    errors_ti = errors.get(ti, None)
+                    if errors_ti is None:
+                        errors_ti = OrderedDict()
+                        errors[ti] = errors_ti
```

Comment:
As before, line 387 and 388 can be combined as `errors_ti = errors.setdefault(ti, OrderedDict())`.

Reply:
True. In my opinion, though, it's easier/clearer to read as two lines. `setdefault` implies that it could be already set, whereas inside this `if` block we know it's not.

Originally, I was doing this whole `if` block with `setdefault` instead (i.e., replacing lines 385-388 with `errors_ti = errors.setdefault(ti, OrderedDict())`). Then I realized that that makes the `OrderedDict` every time, even if it's not used (i.e. even if `ti in errors`), which is wasteful. So that's why I made the `if` block.